minica/minica_test.go

```diff
@@ -0,0 +1,541 @@
+				}); err != nil {
+					t.Errorf("Certificate.Verify() error = %v", err)
+				}
+			}
+		})
+	}
+}
+
+func TestCA_SignSSH(t *testing.T) {
+	signer, err := keyutil.GenerateDefaultSigner()
+	if err != nil {
+		t.Fatal(err)
+	}
+	publicKey, err := ssh.NewPublicKey(signer.Public())
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	type args struct {
+		cert *ssh.Certificate
+	}
+	tests := []struct {
+		name          string
+		ca            *CA
+		args          args
+		wantCertType  uint32
+		wantPrincipal string
+		wantErr       bool
+	}{
+		{"ok host", mustCA(t), args{&ssh.Certificate{
+			Key:             publicKey,
+			Serial:          1234,
+			CertType:        ssh.HostCert,
+			KeyId:           "ssh.test.com",
+			ValidPrincipals: []string{"ssh.test.com"},
+		}}, ssh.HostCert, "ssh.test.com", false},
+		{"ok user", mustCA(t), args{&ssh.Certificate{
+			Key:             publicKey,
+			Serial:          1234,
+			CertType:        ssh.UserCert,
+			KeyId:           "jane@test.com",
+			ValidPrincipals: []string{"jane"},
+			ValidAfter:      uint64(time.Now().Unix()),
+			ValidBefore:     uint64(time.Now().Add(time.Hour).Unix()),
+		}}, ssh.UserCert, "jane", false},
+		{"ok infinity", mustCA(t), args{&ssh.Certificate{
+			Key:             publicKey,
+			Serial:          1234,
+			CertType:        ssh.UserCert,
+			KeyId:           "jane@test.com",
+			ValidPrincipals: []string{"jane"},
+			ValidBefore:     ssh.CertTimeInfinity,
+		}}, ssh.UserCert, "jane", false},
+		{"fail type", mustCA(t), args{&ssh.Certificate{
+			Key:             publicKey,
+			Serial:          1234,
+			CertType:        100,
+			KeyId:           "jane@test.com",
+			ValidPrincipals: []string{"jane"},
+		}}, 0, "", true},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			got, err := tt.ca.SignSSH(tt.args.cert)
+			if (err != nil) != tt.wantErr {
+				t.Errorf("CA.SignSSH() error = %v, wantErr %v", err, tt.wantErr)
+				return
+			}
+
+			if tt.wantErr {
+				if got != nil {
+					t.Errorf("CA.SignSSH() = %v, want nil", got)
+				}
+			} else {
+				checker := ssh.CertChecker{
+					IsUserAuthority: func(auth ssh.PublicKey) bool {
+						return reflect.DeepEqual(auth, tt.ca.SSHUserSigner.PublicKey())
+					},
+					IsHostAuthority: func(auth ssh.PublicKey, address string) bool {
+						return reflect.DeepEqual(auth, tt.ca.SSHHostSigner.PublicKey())
+					},
+				}
+				switch tt.wantCertType {
+				case ssh.HostCert:
+					if err := checker.CheckHostKey(tt.wantPrincipal+":22", &net.IPAddr{IP: net.IP{1, 2, 3, 4}}, got); err != nil {
+						t.Errorf("CertChecker.CheckHostKey() error = %v", err)
+					}
+				case ssh.UserCert:
+					if _, err := checker.Authenticate(mockConnMetadata(tt.wantPrincipal), got); err != nil {
+						t.Errorf("CertChecker.Authenticate() error = %v", err)
+					}
+				default:
+					t.Fatalf("unknown cert type %v", tt.wantCertType)
+				}
+			}
+		})
+	}
+}
+
+func TestCA_SignSSH_mutation(t *testing.T) {
```

Comment:
I see a naming pattern with underscores for tests. Not quite sure why the linter isn't complaining but maybe `TestCASignSSHMutation` should be a more idiomatic name? This would apply to the rest as well.

Reply:
A well-known tool named [gotests](https://github.com/cweill/gotests) defaults to `TestName` for functions and `TestStruct_Name` for methods. I'm adding another `_` when I want to add a test for a special use case, so it doesn't look like another method.